### PR TITLE
fix: implement a retry utility and wire it into the Fal billing lookup

### PR DESCRIPTION
**Pull request by gianpaj**
- `apps/web/lib/fetch-with-retry.ts`: reusable `fetchWithRetry<T>()` with configurable delays, timeout, request options, and response parsing. Defaults to **three retries after the first attempt, waiting 1s/2s/4s**, with a fresh 5-second timeout each time.
- `apps/web/lib/fal-billing.ts`: retries missing or invalid cost data, HTTP failures, and network errors. Returns immediately on success, including a zero cost. After exhaustion, logs **one warning** and returns `null`.
- `apps/web/app/api/clone-voice/route.ts`: uses the extracted helper and preserves the estimated-cost fallback.
- Added tests and updated `docs/devops.md`.

To fix error: `Fal billing events API returned unexpected cost data`

### CODE REVIEW

**Comment by gianpaj** [comment]
re: Zero-cost billing events collapse into `basis: 'unknown'`


Then this is inert for completed Fal enhancements, and I’d leave the code unchanged.

The zero-cost test checks a defensive edge case, not an observed Fal response. It doesn’t justify changing retries or downstream cost accounting. If Fal starts returning `0`, we should confirm its meaning before deciding whether to retry it or record it as a genuine free charge.

File: apps/web/lib/fetch-with-retry.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  for (let attempt = 0; ; attempt++) {
+    try {
+      const response = await fetch(url, {
+        ...requestInit,
+        signal: AbortSignal.timeout(timeoutMs),
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
+      }
```
[the remark this answers is not in the document]
Comment (gianpaj) [reply apps/web/lib/fetch-with-retry.ts:52]: Zed - GPT-6-Astra

Addressed in d179f2ca. HTTP 400/401/403/404 and other non-retryable statuses fail immediately. Only 408, 429, and 5xx responses retry. A valid Retry-After header overrides the scheduled delay, supporting seconds and HTTP dates; missing or invalid headers fall back to 1s/2s/4s. The three-retry limit remains. Fal logs once and returns null on terminal failure.

Added tests for permanent errors, retryable statuses, Retry-After timing and fallback, and retry exhaustion. Updated the operational docs.

Validation: pnpm fixall, pnpm type-check, and all 109 focused tests passed. The full suite had 1,092 passing tests and the same 35 Stripe webhook failures caused by the Redis connection issue.

File: apps/web/tests/fetch-with-retry.test.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      await vi.advanceTimersByTimeAsync(1);
+      expect(fetchMock).toHaveBeenCalledTimes(index + 2);
+    }
+    await result;
+    expect(vi.getTimerCount()).toBe(0);
+  });
+
+  it('retries HTTP errors and stops after a successful retry', async () => {
+    fetchMock
+      .mockResolvedValueOnce(new Response(null, { status: 503 }))
```
[the remark this answers is not in the document]
Comment (gianpaj) [reply apps/web/tests/fetch-with-retry.test.ts]: Zed - GPT-6-Astra

Addressed in 1af697e0. The 503 response has a valid JSON body, so parsing cannot accidentally drive the retry. Without the HTTP status guard, the helper would return the error object and fail the assertion.

Validation: all 109 focused tests across fetch-with-retry, fal-billing, and clone-voice passed.

File: apps/web/lib/fetch-with-retry.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  const seconds = Number(value);
+  if (Number.isFinite(seconds)) {
+    return seconds >= 0 ? seconds * 1000 : undefined;
+  }
+
+  const date = Date.parse(value);
+  return Number.isFinite(date) ? Math.max(0, date - Date.now()) : undefined;
+}
```
[the remark this answers is not in the document]
Comment (gianpaj) [reply apps/web/lib/fetch-with-retry.ts:24]: Zed - GPT-6-Astra

Addressed in c8304ac0. The helper has a configurable maxTotalDelayMs budget, defaulting to 30 seconds across all retry sleeps, including fallback delays. If the next delay exceeds the remaining budget, it throws the last error immediately rather than clamping the delay and retrying before the server permits it. The Fal caller catches that error, logs once, and returns null so estimated-cost tracking and analytics can continue.

Numeric Retry-After values must contain integer seconds. Values such as 1e3 and 1.5 use the fallback schedule. Huge values, far-future dates, non-finite delays, and timer-overflow values cannot bypass the sleep budget.

Updated docs/devops.md to state the default 50-second budget: at most 20 seconds of request timeouts plus 30 seconds of scheduled sleeps, excluding local processing and event-loop scheduling. This bounds retry sleeps, not the entire caller invocation.

Validation: pnpm fixall, pnpm type-check, git diff --check, and all 123 focused tests passed. Tests cover cumulative sleep, the exact budget boundary, oversized headers, fallback delays sharing the budget, and Fal returning null without waiting on an over-budget header.
